lib/bidding-strategies/queuepending.js

```diff
@@ -0,0 +1,142 @@
+   * the constructor arguments are always JSON-serializable and can be passed
+   * directly from the service configuration file (e.g. config.yml).
+   *
+   * For unit testing with a different queue object, the property `_queue`
+   * (e.g. instance._queue) will be defined and supported as a
+   * taskcluster-client Queue object which can be substituted using Sinon or
+   * similar purpose libraries.
+   */
+  constructor({rootUrl, credentials}) {
+    super({id: 'queue-pending-rootUrl'});
+    this._queue = new taskcluster.Queue({rootUrl, credentials});
+  }
+
+  /**
+   * Internal utility method wrapping the actual queue call for easy
+   * mocking in unit tests
+   */
+  async _getPending(workerType) {
+    let {pendingTasks} = await this._queue.pendingTasks('worker-manager', workerType);
+    return pendingTasks;
+  }
+
+  /**
+   * Given a workerType and biddingStrategyData, determine how much 
+   * demand there is for this workerType.
+   */
+  async determineDemand({workerType, biddingStrategyData}, runningCapacity, pendingCapacity) {
+    let {
+      scalingRatio,
+      minCapacity,
+      maxCapacity,
+      minPrice,
+      maxPrice,
+    } = biddingStrategyData;
+    
+    // TODO: we need to pass in pending and running capacity into all bidding
+    // strategies so that they may calculate demenand based on overall capacity
+    // available and planned without having to talk to providers
+
+    let pending = await this._getPending(workerType);
+
+    // NOTE: The following code of this function was copied verbatim from the
+    // aws-provisoner-v1 codebase, changed only to address different variable
+    // scoping (e.g. removing `this.` for property lookups and removing logging
+    // which is not set up in this repository at the time of writing
+
+    // scalingRatio = 0.2 => keep pending tasks as 20% of runningCapacity
+    // scalingRatio = 0   => keep pending tasks as  0% of runningCapacity
+    let desiredPending = Math.round(scalingRatio * runningCapacity);
+
+    // desiredPending < pending - pendingCapacity    =>   Create spot requests
+    // otherwise Cancel spot requests
+    let capacityChange = pending - pendingCapacity - desiredPending;
+
+    // capacityChange > 0  => Create spot requests for capacityChange
+    // capacityChange < 0  => cancel spot requests for capacityChange
+    let capacityAfterChange = capacityChange + pendingCapacity + runningCapacity;
+
+    if (capacityAfterChange >= maxCapacity) {
+      // If there is more than max capacity we should always aim for maxCapacity
+      return maxCapacity - runningCapacity - pendingCapacity;
+    } else if (capacityAfterChange < minCapacity) {
+      return minCapacity - runningCapacity - pendingCapacity;
+    }
+
+    // If we're not hitting limits, we should aim for the capacity change that
+    // fits with the scalingRatio, to keep pending tasks around as a percentage
+    // of the running capacity.
+    return capacityChange;
+  }
+
+  /**
```

Comment:
From here down looks fairly generic, and not specific to the QueuePending strategy.  Would it make sense to move these implementations to the parent class?  That could be a later change, obviously -- they just struck me as out of place so I thought I'd mention it.

Reply:
I see why you're thinking that.  I didn't want to force the assumption that this was the best way to select bids, since it's basically a price based comparison.  I think that if we have a second (or more) bidding strategy and those bidding strategies are just copying this method, we should consider moving this to the base class.  I'd rather not make that change right now though, especially as I suspect we'll only have this one bidding strategy for some time.